### PR TITLE
feat(unicommerce): Sales Invoice creation through picklist

### PR DESCRIPTION
**Pull request by sonali8848**
- Added one child table of order details in Pick List to get orders and Generate Invoice button at the top. 
image

- Once the picklist is saved, the sales orders are appended to the Order details child table in the pick list.
image

- Once the sales invoice is generated, in the Order details table in the picklist, the sales invoice no. and pdf of the invoice are attached in the pick list.
image

### CODE REVIEW

File: ecommerce_integrations/fixtures/custom_field.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
+[
+ {
+  "allow_in_quick_entry": 0,
```
Comment (ankush) [inline ecommerce_integrations/fixtures/custom_field.json:3]: Custom field in this app should be managed by settings doctype's setup_custom_field function

https://github.com/frappe/ecommerce_integrations/blob/5cafd9129edcd4951fff4a51117e6bc58c91cea6/ecommerce_integrations/unicommerce/doctype/unicommerce_settings/unicommerce_settings.py#L189

Comment (ankush) [reply ecommerce_integrations/fixtures/custom_field.json:3]: This is required to not automatically add custom fields which might not be required by everyone.

File: ecommerce_integrations/public/js/unicommerce/pick_list.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                                        }
+                                      else{
+                                          return {}
+                                      }
+                        });
+                        sales_orders.push(so);
+                            warehouse_allocation[so] = item_details.filter(value => Object.keys(value).length !== 0);
+                    });
+                    frappe.call({
+			method: 'ecommerce_integrations.unicommerce.pick_list.validate_partial_picking',
```
Comment (ankush) [inline ecommerce_integrations/public/js/unicommerce/pick_list.js:37]: These validations should ideally be done on server side only?

File: ecommerce_integrations/unicommerce/inventory.py [hunk 3]
```diff
@@ -86,3 +90,65 @@ def _update_inventory_sync_status(ecom_item_success_map: Dict[str, bool], timest
...
+	wh_to_facility_map = settings.get_erpnext_to_integration_wh_mapping()
+	facility_code = wh_to_facility_map[warehouse]
+	inventory_list = []
+	for shelf in shelves:
+		inventoryType = "GOOD_INVENTORY"
+		if shelf.type == "Unsellable":
+			inventoryType = "BAD_INVENTORY"
+		for item in erpnext_inventory:
+			custom_filter = {
+				"company": "Lifelong Online Retail Private Limited",
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/inventory.py:120]: What?

File: ecommerce_integrations/unicommerce/inventory.py [hunk 4]
```diff
@@ -86,3 +90,65 @@ def _update_inventory_sync_status(ecom_item_success_map: Dict[str, bool], timest
 	for ecom_item, status in ecom_item_success_map.items():
 		if status:
 			update_inventory_sync_status(ecom_item, timestamp)
+
+
+def shelf_bulk_update(warehouse, settings):
+	warehouse = warehouse["erpnext_warehouse"]
+	shelves = frappe.get_list("Shelf", {"warehouse": warehouse, "disable": 0}, ["shelf_name", "type"])
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/inventory.py:97]: This appears to be a custom doctype. Why is it added in core here?

File: ecommerce_integrations/unicommerce/inventory.py [hunk 5]
```diff
@@ -86,3 +90,65 @@ def _update_inventory_sync_status(ecom_item_success_map: Dict[str, bool], timest
...
+				"item_code": item.item_code,
+			}
+			columns, data = report.get_data(limit=1, filters=custom_filter, as_dict=True)
+			if len(data) > 0:
+				data = data[0]
+				inventory_list.append(
+					{
+						"itemSKU": item.item_code,
+						"quantity": data["qty_after_transaction"],
+						"shelfCode": shelf,
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/inventory.py:134]: Here underlying assumption is that shelf name on unicommerce is same as what it is on ERPNext. It should instead be a mapping like warehouse->location.

File: ecommerce_integrations/unicommerce/pick_list.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+			"You are picking more than required quantity for "
+			+ so_item_doc.item_code
+			+ ". Check if there is any other picklist for "
+			+ so_item_doc.parent
+		)
+
+
+@frappe.whitelist()
+def validate(self, method=None):
+	if self.get("locations"):
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/pick_list.py:25]: Can you please briefly describe what is going on here? quite large function that's validating and setting lot of things.

**Review by ankush — CHANGES_REQUESTED** [review]
Some custom development appear to be part of this PR, if you intend to add something to core apps you should consider making it generic/configurable always :sweat_smile:

**Comment by ankush** [comment]
@sonali8848 are all review comments addressed? mark as ready for review and drop replies or mark them as resolved.

**Comment by sonali8848** [comment]
Sure

File: ecommerce_integrations/public/js/unicommerce/pick_list.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                                        }
+                                })
+                     frappe.call({
+                                method: 'ecommerce_integrations.unicommerce.invoice.generate_unicommerce_invoices',
+                                args: {
+                                        'sales_orders': sales_orders,
+                                        'warehouse_allocation': warehouse_allocation
+                                        },
+                                freeze: true,
+                                freeze_message: "Requesting Invoice generation. Once synced, invoice will appear in linked documents.",
```
[the remark this answers is not in the document]
Comment (sonali8848) [reply ecommerce_integrations/public/js/unicommerce/pick_list.js:43]: Done

File: ecommerce_integrations/unicommerce/invoice.py [hunk 8]
```diff
@@ -570,3 +570,57 @@ def update_cancellation_status(so_data, so) -> bool:
 	from ecommerce_integrations.unicommerce.cancellation_and_returns import update_erpnext_order_items
 
 	update_erpnext_order_items(so_data, so)
+
+
+@frappe.whitelist()
```
[the remark this answers is not in the document]
Comment (sonali8848) [reply ecommerce_integrations/unicommerce/invoice.py]: Done

File: ecommerce_integrations/unicommerce/invoice.py [hunk 9]
```diff
@@ -570,3 +570,57 @@ def update_cancellation_status(so_data, so) -> bool:
...
+		pl_doc = frappe.get_doc("Pick List", pl.parent)
+		for i in pl_doc.get("order_details"):
+			if not i.sales_invoice:
+				is_invoice_generated = 0
+				break
+		if is_invoice_generated:
+			frappe.db.set_value("Pick List", pl.parent, "workflow_state", "Invoice Generated")
+
+
+@frappe.whitelist()
```
[the remark this answers is not in the document]
Comment (sonali8848) [reply ecommerce_integrations/unicommerce/invoice.py]: Resolved

File: ecommerce_integrations/unicommerce/pick_list.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	):
+		frappe.throw(
+			"You are picking more than required quantity for "
+			+ so_item_doc.item_code
+			+ ". Check if there is any other picklist for "
+			+ so_item_doc.parent
+		)
+
+
+@frappe.whitelist()
```
[the remark this answers is not in the document]
Comment (sonali8848) [reply ecommerce_integrations/unicommerce/pick_list.py]: Resolved

File: ecommerce_integrations/unicommerce/pick_list.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+						+ str(pl.idx)
+						+ ". Pick the item to proceed!"
+					)
+		item_so_list = [d.sales_order for d in self.get("locations")]
+		unique_so_list = []
+		for i in item_so_list:
+			if i not in unique_so_list:
+				unique_so_list.append(i)
+		if len(unique_so_list) > 30:
+			frappe.throw("Cannot add more than 30 Sales Orders!")
```
[the remark this answers is not in the document]
Comment (sonali8848) [reply ecommerce_integrations/unicommerce/pick_list.py]: Resolved

File: ecommerce_integrations/public/js/unicommerce/pick_list.js (hunk 2 above)
Comment (sonali8848) [reply ecommerce_integrations/public/js/unicommerce/pick_list.js]: Done

File: ecommerce_integrations/unicommerce/inventory.py (hunk 3 above)
Comment (sonali8848) [reply ecommerce_integrations/unicommerce/inventory.py]: Its is added by mistake, Its removed.

File: ecommerce_integrations/unicommerce/inventory.py (hunk 5 above)
Comment (sonali8848) [reply ecommerce_integrations/unicommerce/inventory.py]: Removed

File: ecommerce_integrations/unicommerce/doctype/unicommerce_settings/unicommerce_settings.py [hunk 12]
```diff
@@ -427,6 +428,15 @@ def setup_custom_fields(update=True):
 				read_only=1,
 			),
 		],
+		"Pick List": [
+			dict(
+				fieldname=ORDER_DETAILS,
+				label="Order Details",
+				fieldtype="Table",
+				insert_after="section_break_15",
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/doctype/unicommerce_settings/unicommerce_settings.py:436]:
```suggestion
```

Remove this, it's unreliable. Putting table at end should be fine.

File: ecommerce_integrations/unicommerce/invoice.py [hunk 13]
```diff
@@ -570,3 +570,47 @@ def update_cancellation_status(so_data, so) -> bool:
 	from ecommerce_integrations.unicommerce.cancellation_and_returns import update_erpnext_order_items
 
 	update_erpnext_order_items(so_data, so)
+
+
+def on_submit(self, method=None):
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/invoice.py:575]: This hook will start triggering on invoices that aren't generated by unicommerce at all.


Make this code a bit more resilient? 


Consider all these use cases:
1. The invoice is not even generated by Unicommerce, should this hook even run? 
2. The invoice is generated via unicommerce but there's no picklist created for it, it's directly generated or manually generated.

File: ecommerce_integrations/unicommerce/invoice.py [hunk 14]
```diff
@@ -570,3 +570,47 @@ def update_cancellation_status(so_data, so) -> bool:
...
+		else:
+			frappe.db.set_value("Pick List Sales Order Details", pl.name, {"sales_invoice": self.name})
+
+
+def on_cancel(self, method=None):
+	results = frappe.db.get_all(
+		"Pick List Sales Order Details", filters={"sales_invoice": self.name, "docstatus": 1}
+	)
+	if results:
+		self.flags.ignore_links = True
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/invoice.py:616]: Why is this added? This will ignore ALL links and cause data inconsistency. 

If you want to ignore Pick List links be more explicit. 


```python
ignored_doctypes = list(self.get("ignore_linked_doctypes", []))
ignored_doctypes.append("Pick List")
self.ignore_linked_doctypes = ignored_doctypes
```   


I haven't tested this but it should work. Test it out once.

File: ecommerce_integrations/unicommerce/pick_list.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
+import json
+
+import frappe
+from frappe import _
+
+
+def validate(self, method=None):
```
Comment (ankush) [inline ecommerce_integrations/unicommerce/pick_list.py:7]: This validation should only apply if Pick List is generated for unicommerce orders? 


Similar to my comments on Sales Invoice, make sure that other use cases don't break while validating for unicommerce.

**Review by ankush — CHANGES_REQUESTED** [review]

**Comment by ankush** [comment]
@sonali8848 I've reviewed again, last few changes pending after which it should be mergeable :+1: 


PS: In future please raise PR from personal forks or grant me access to 8848's fork otherwise I can't make changes to PRs.

**Review by ankush — APPROVED** [review]
